File: wear/src/main/java/com/atakmap/android/wickr/plugin/multisensortracking/SpO2Listener.java
/*
 * Copyright 2022 Samsung Electronics Co., Ltd. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.atakmap.android.wickr.plugin.multisensortracking;

import android.util.Log;

import androidx.annotation.NonNull;

import com.atakmap.android.wickr.plugin.R;
import com.samsung.android.service.health.tracking.HealthTracker;
import com.samsung.android.service.health.tracking.data.DataPoint;
import com.samsung.android.service.health.tracking.data.ValueKey;

import java.util.List;

public class SpO2Listener extends BaseListener {
    private final static String APP_TAG = "SpO2Listener";

    SpO2Listener() {
        final HealthTracker.TrackerEventListener trackerEventListener = new HealthTracker.TrackerEventListener() {
            @Override
            public void onDataReceived(@NonNull List<DataPoint> list) {
                for (DataPoint data : list) {
                    updateSpo2(data);
                }
            }

            @Override
            public void onFlushCompleted() {
                Log.i(APP_TAG, " onFlushCompleted called");
            }

            @Override
            public void onError(HealthTracker.TrackerError trackerError) {
                Log.e(APP_TAG, " onError called: " + trackerError);
                setHandlerRunning(false);
                if (trackerError == HealthTracker.TrackerError.PERMISSION_ERROR) {
                    TrackerDataNotifier.getInstance().notifyError(R.string.NoPermission);
                }
                if (trackerError == HealthTracker.TrackerError.SDK_POLICY_ERROR) {
                    TrackerDataNotifier.getInstance().notifyError(R.string.SdkPolicyError);
                }
            }
        };
        setTrackerEventListener(trackerEventListener);
    }

    public void updateSpo2(DataPoint dataPoint) {
        final int status = dataPoint.getValue(ValueKey.SpO2Set.STATUS);
        int spo2Value = 0;
        if (status == SpO2Status.MEASUREMENT_COMPLETED)
            spo2Value = dataPoint.getValue(ValueKey.SpO2Set.SPO2);
        TrackerDataNotifier.getInstance().notifySpO2TrackerObservers(status, spo2Value);
        Log.d(APP_TAG, dataPoint.toString());
    }

}

File: wear/src/main/java/com/atakmap/android/wickr/plugin/multisensortracking/SpO2Listener.kt
package com.atakmap.android.wickr.plugin.multisensortracking

import com.atakmap.android.wickr.plugin.R
import com.samsung.android.service.health.tracking.HealthTracker.TrackerError
import com.samsung.android.service.health.tracking.HealthTracker.TrackerEventListener
import com.samsung.android.service.health.tracking.data.DataPoint
import com.samsung.android.service.health.tracking.data.ValueKey
import org.koin.core.component.KoinComponent
import org.koin.core.component.get

class SpO2Listener internal constructor() : BaseListener(), KoinComponent {

    private val trackerDataNotifier: TrackerDataNotifier = get()

    init {
        val trackerEventListener: TrackerEventListener = object : TrackerEventListener {
            override fun onDataReceived(list: List<DataPoint>) {
                for (data in list) {
                    updateSpo2(data)
                }
            }

            override fun onFlushCompleted() {

            }

            override fun onError(trackerError: TrackerError) {
                setHandlerRunning(false)
                if (trackerError == TrackerError.PERMISSION_ERROR) {
                    trackerDataNotifier.notifyError(R.string.NoPermission)
                }
                if (trackerError == TrackerError.SDK_POLICY_ERROR) {
                    trackerDataNotifier.notifyError(R.string.SdkPolicyError)
                }
            }
        }

        setTrackerEventListener(trackerEventListener)
    }

    fun updateSpo2(dataPoint: DataPoint) {
        val status = dataPoint.getValue(ValueKey.SpO2Set.STATUS)
        var spo2Value = 0
        if (status == SpO2Status.MEASUREMENT_COMPLETED) spo2Value =
            dataPoint.getValue(ValueKey.SpO2Set.SPO2)
        trackerDataNotifier.notifySpO2TrackerObservers(status, spo2Value)
    }
}
